Exercise 2

muzaluisa opened the pull request:
Hi Siim,

Taking into account pylint, unittesting and logging recommendations, I have made exercise 2 (in 1 file):)
I am still not sure, what is a good practice for docstring, could you give me some hints?

muzaluisa committed:
Exercise 2

siimphh commented on codeU2.py:
```diff
@@ -0,0 +1,210 @@
+import logging
+import unittest
+
+def dfs(tree, start):
```
Is this function unused? Sometimes unused functions may be left behind but generally code is not added if it is not used or not planned to be used.

muzaluisa replied:
Unused indeed, I have removed that :)

siimphh commented on codeU2.py:
```diff
@@ -0,0 +1,210 @@
+
+def dfs_path(tree, start, end, path=None):
+
+    """Returns all full paths from start node
```
Does this return multiple paths because the same element could be in the tree multiple times?

Also, it is might be clearer to mention the function "Yields" all full path rather than "Returns" as that distinction is important for the caller (requiring the explicit `list()` in the example in this file).

muzaluisa replied:
The function works for the cases, when the input is a graph as well or the element indeed occurs multiple times, that is why multiple paths are assumed. Added the comment :)

siimphh commented on codeU2.py:
```diff
@@ -0,0 +1,210 @@
+    to end node, including the end node
+    """
+
+    stack = [(start, [start])]
```
When defining non-trivial structures, it is helpful to explain them to the readers. It takes some effort trying to figure out how the nested lists and tuples are supposed to work in the loop below.

siimphh commented on codeU2.py:
```diff
@@ -0,0 +1,210 @@
+            else:
+                stack.append((next, path+[next]))
+
+def get_head(tree):
```
It sounds like determining the head is something users would want to do frequently for a tree. Would it be worth storing it in an attribute?

muzaluisa replied:
Right I have the class for storing all the information about the tree:
class Tree():
    
    """Defines the class for the tree:
    self.children contains a dict of pairs (node:(left_child,right_child))
    self.head contains the value of the tree head
    None denotes the absence if the node
    """
    
    def __init__(self,children,head):
        self.children = children;
        self.head = head;

siimphh commented on codeU2.py:
```diff
@@ -0,0 +1,210 @@
+    if not tree:
+        return None
+    all_nodes = set()
+    for node in tree.keys():
```
Would it be more to the point to iterate over `tree.values()`?

muzaluisa replied:
True, that is more logical :)

muzaluisa replied:
I added head attribute, so the function itself is not needed

siimphh commented on codeU2.py:
```diff
@@ -0,0 +1,210 @@
+    for node in tree.keys():
+        if node:
+            all_nodes.update(list(tree[node]))
+    head = (set(tree.keys())-all_nodes-set([None])).pop()
```
I'm confused why a key would be `None`, is there a particular reason to support that?

muzaluisa replied:
Since left or right child of the node could be empty, and I need to indicate by None.

siimphh commented on codeU2.py:
```diff
@@ -0,0 +1,210 @@
+
+    if not tree:
+        return None
+    all_nodes = set()
```
Would it make it slightly clearer to rename `all_nodes` to `all_children` and `set(tree.keys())` (below in the final expression) `all_parents`?

I'm specifically looking at the set arithmetic expression below where it is not immediately obvious why `all_nodes` and `tree.keys()` are subtracted as they are and thinking that a little bit of naming might help.

muzaluisa replied:
You are right, I need to take more precise names of the variables. Fixed)

siimphh commented on codeU2.py:
```diff
@@ -0,0 +1,210 @@
+    if head == node:
+        logging.warning('Head has no ancestors')
+        return None
+    ancestors = list(dfs_path(tree, head, node, path=None))
```
It looks a bit awkward to have to explicitly `list()` the return value from the generator and then also following it up with `ancestors[0][:-1]` below. It might not be appropriate to change the `dfs_path()` return value but I'd suggest at least considering it if a fair amount of postprocessing is required in the caller.

muzaluisa replied:
True, now I return path excluding the end node, so taking ancestors[0] is only used.
Does that makes more sense now or smth else needs to be done?

def dfs_path(tree_, start, end, path=None):

    """Yields all full paths from start node
    to end node, excluding the end node
    """
    
    tree = tree_.children
    stack = [(start, [start])]
    while stack:
        (vertex, path) = stack.pop()
        if vertex is None:
            continue
        for next in set(tree[vertex]) - set(path):
            if next == end:
                yield path

siimphh reviewed (commented):
I felt that in a few places either the code could have used clarifications either by more indicative variable naming, comments or better yet - simplifying the entire approach.

The algorithmic approach looks entirely feasible. And I appreciate that the second assignment already used more "idiomatic" Python code.

siimphh commented:
Docstring practices differ from code base to code base depending on static analysis tools and documentation generation tools in use. For example, Google Python style guide has this to say:

http://google.github.io/styleguide/pyguide.html?showone=Comments#Comments

muzaluisa committed:
Taking into account Siim's comments

1. Added Tree class
2. Fixed dfs_path so that it returns the path from node to the end node excluding it
3. Fixed all other small comments

muzaluisa commented:
Hi, Siim!

Thank you for the feedback :) Could you have a look at updated version?

siimphh commented on codeU2.py:
```diff
@@ -7,102 +7,73 @@
-    return visited
-
-def dfs_path(tree, start, end, path=None):
+def dfs_path(tree_, start, end, path=None):
```
Since these functions take a `Tree` as the first argument, It looks like it would be appropriate to make them be methods:

```
class Tree
  ...
  def dfs_path(self, start, end, path=None):
    ...
```

siimphh commented on codeU2.py:
```diff
@@ -0,0 +1,181 @@
+    to end node, excluding the end node
+    """
+    
+    tree = tree_.children
```
This rename makes it a bit more confusing for me. Below the meaning of `tree[vertex]` is not entirely obvious (though one might guess) whereas `tree.children[vertex]` is less ambiguous.

siimphh commented on codeU2.py:
```diff
@@ -0,0 +1,181 @@
+    """
+    
+    tree = tree_.children
+    stack = [(start, [start])]
```
I'd like to see a little comment explaining the algorithm implemented here. It looks like descending depth first into the tree breaking the descent (and trying other branches) when the `end` node is found?

The algorithm explicitly allows for `end` values in the tree (by not breaking on the first path found) but then the set arithmetic `set(tree[vertex]) - set(path)` looks inconsistent in not allowing multiples of any intermediate nodes.

Is that to not get into loops? One would imagine that loops should not be allowed (it wouldn't be a tree if they were). Maybe it would be better to check for loops in the constructor. For the purposes of this assignment, it would probably be enough to just document that "trees" with loops are forbidden and leave the actual check out.

In any case, it took me a bit of time to trace what this function is doing and I suspect it took you a bit of careful thinking to write it as well. Having to carefully think about your code is a great signal for you to also write some comments to the code so the future readers can skip the "figuring out" part and get the conclusions from the helpful comments.

siimphh commented on codeU2.py:
```diff
@@ -0,0 +1,181 @@
+        self.children = children;
+        self.head = head;
+
+def dfs_path(tree_, start, end, path=None):
```
The `path=None` is never overridden anywhere so the argument should be removed for simplicity. If this was a `Tree` class method then `tree_` would be `self` and `start` could be omitted for `self.head` as well.

siimphh replied:
In fact, I think all of these functions should be methods on `Tree`:

```
class Tree:
  def __init__(self, ...):
    ...

  def dfs_path(self, ...):
    ...

  def get_ancestores(self, ...):
    ...
...
```

siimphh commented on codeU2.py:
```diff
@@ -0,0 +1,181 @@
+    """
+
+    head = tree_.head
+    if not head:
```
Just food for thought - does `dfs_path` handle `head=None` and `head=node` cases? Should it? If these checks are required (functionally) at all, then which function should they belong to?

It's a different story if you want to log these conditions for visibility into how your algorithm works for debugging. In that case of course you can log interesting conditions wherever you need to know of them :)

siimphh commented on codeU2.py:
```diff
@@ -0,0 +1,181 @@
+
+    tree = tree_.children
+    head = tree_.head
+    if not tree:
```
Likewise for these checks. Some of them duplicate checks that are done above (`node1 == head`) and the checking could probably be simplified. Could/should `get_ancestors` return an empty `list` (`[]`) instead of `None` for that case? It sounds like it would make sense and then the algorithm would just work for `node == head` case as well without any extra checks required.

siimphh reviewed (commented):
A few more comments. I'm not sure if I'm doing a good job explaining the background for some of the comments - let me know if more examples or explanation would be useful.
